src/main/java/org/opensearch/neuralsearch/highlight/NeuralHighlighterEngine.java

```diff
@@ -0,0 +1,306 @@
+                        if (item instanceof Map<?, ?> map) {
+                            Map<String, Object> safeMap = new java.util.HashMap<>();
+                            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                                safeMap.put(entry.getKey().toString(), entry.getValue());
+                            }
+
+                            // Extract and validate positions
+                            Object startObj = safeMap.get(MODEL_INFERENCE_RESULT_START_KEY);
+                            Object endObj = safeMap.get(MODEL_INFERENCE_RESULT_END_KEY);
+
+                            int[] positions = validateHighlightPositions(startObj, endObj, context.length());
+                            if (positions != null) {
+                                validHighlights.add(positions);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // If no valid highlights, return the original text
+        if (validHighlights.isEmpty()) {
+            return context;
+        }
+
+        // Sort highlights by start position (ascending)
+        validHighlights.sort(Comparator.comparingInt(pos -> pos[0]));
+
+        return constructHighlightedText(context, validHighlights);
+    }
+
+    /**
+     * Constructs highlighted text by iterating through the text once
+     *
+     * @param text The original text
+     * @param highlights The list of valid highlight positions (sorted by start position)
+     * @return The highlighted text
+     */
+    private String constructHighlightedText(String text, List<int[]> highlights) {
+        StringBuilder result = new StringBuilder();
+        int currentPos = 0;
+
+        // Check for overlapping highlights and merge them if necessary
+        List<int[]> mergedHighlights = mergeOverlappingHighlights(highlights);
+
+        // Iterate through each highlight
+        for (int[] highlight : mergedHighlights) {
+            int start = highlight[0];
+            int end = highlight[1];
+
+            // Add text before the highlight
+            if (start > currentPos) {
+                result.append(text, currentPos, start);
+            }
+
+            // Add the highlighted text
+            result.append(DEFAULT_PRE_TAG);
+            result.append(text, start, end);
+            result.append(DEFAULT_POST_TAG);
+
+            // Update current position
+            currentPos = end;
+        }
+
+        // Add any remaining text
+        if (currentPos < text.length()) {
+            result.append(text, currentPos, text.length());
+        }
+
+        return result.toString();
+    }
+
+    /**
+     * Merges overlapping highlights to ensure proper nesting
+     *
+     * @param highlights The list of highlights sorted by start position
+     * @return A list of merged highlights with no overlaps
+     */
+    private List<int[]> mergeOverlappingHighlights(List<int[]> highlights) {
```

Comment:
Shouldn't model generate non overlapping highlight?

Reply:
Right, both our provided model and other models should handle this logic on server side, removed this part for latency conern.